File: scripts/notif-poller.ts
#!/usr/bin/env npx tsx
/**
 * macOS Notification → Streamscapes forwarder.
 *
 * Polls the macOS notification center database and forwards new
 * notifications to the Streamscapes ingest endpoint as spans.
 *
 * Requires macOS. Your terminal needs Full Disk Access:
 *   System Settings → Privacy & Security → Full Disk Access → add Terminal/iTerm
 *
 * Usage:
 *   npx tsx --env-file .env scripts/notif-poller.ts          # local
 *   npx tsx --env-file .env scripts/notif-poller.ts --remote  # live
 */

import { execSync } from 'child_process';
import { homedir } from 'os';
import { join } from 'path';

const useRemote = process.argv.includes('--remote') || process.argv.includes('-r');
const SS_API_KEY = useRemote ? (process.env.SS_PROD_API_KEY ?? process.env.SS_API_KEY) : (process.env.SS_LOCAL_API_KEY ?? process.env.SS_API_KEY);
const SS_PROD_ENDPOINT = process.env.SS_PROD_ENDPOINT ?? 'https://www.streamscapes.fm/api/ingest/otlp/v1/traces';
const SS_ENDPOINT = process.env.SS_ENDPOINT ?? (useRemote ? SS_PROD_ENDPOINT : 'http://localhost:3000/api/ingest/otlp/v1/traces');
const POLL_INTERVAL_S = Number(process.env.NOTIF_POLL_INTERVAL_S) || 3;

if (!SS_API_KEY) {
  console.error('Missing SS_API_KEY — generate one in the Streamscapes Connections panel');
  process.exit(1);
}

const DB_PATH = join(homedir(), 'Library/Group Containers/group.com.apple.usernoted/db2/db');
/** Seconds between Unix epoch (1970) and Mac/Core Data epoch (2001-01-01) */
const MAC_EPOCH_OFFSET = 978307200;

const seenIds = new Set<number>();
// Start from "now" in Mac epoch seconds
let lastDeliveredDate = Date.now() / 1000 - MAC_EPOCH_OFFSET;

interface NotifRow {
  rec_id: number;
  identifier: string;
  delivered_date: number;
}

function queryNewNotifications(): NotifRow[] {
  try {
    const query = `SELECT r.rec_id, a.identifier, r.delivered_date FROM record r JOIN app a ON r.app_id = a.app_id WHERE r.delivered_date > ${lastDeliveredDate} ORDER BY r.delivered_date ASC LIMIT 50;`;
    const result = execSync(`sqlite3 -json "${DB_PATH}" "${query}"`, {
      encoding: 'utf-8',
      timeout: 5000,
    }).trim();
    if (!result) return [];
    return JSON.parse(result) as NotifRow[];
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    if (msg.includes('authorization denied') || msg.includes('not permitted') || msg.includes('unable to open')) {
      console.error('[notif] Cannot read notification database.');
      console.error('        Grant Full Disk Access: System Settings → Privacy & Security → Full Disk Access → add your terminal');
      process.exit(1);
    }
    console.warn('[notif] Query error:', msg.slice(0, 200));
    return [];
  }
}

/** Extract notification title/body from the binary plist blob */
function getNotifContent(recId: number): { title?: string; body?: string; subtitle?: string } {
  try {
    // Write blob to temp file, then use `plutil -p` (handles types JSON can't)
    execSync(
      `sqlite3 "${DB_PATH}" "SELECT writefile('/tmp/ss_notif.plist', data) FROM record WHERE rec_id = ${recId};"`,
      { encoding: 'utf-8', timeout: 5000 }
    );
    const raw = execSync('plutil -p /tmp/ss_notif.plist', {
      encoding: 'utf-8',
      timeout: 5000,
    });

    // Parse plutil -p output: "key" => "value" format
    const extract = (key: string): string | undefined => {
      const re = new RegExp(`"${key}"\\s*=>\\s*"([^"]*)"`, 'm');
      const m = raw.match(re);
      return m?.[1] || undefined;
    };

    return {
      title: extract('titl'),
      subtitle: extract('subt'),
      body: extract('body'),
    };
  } catch {
    return {};
  }
}

/** Turn a bundle ID like "com.apple.MobileSMS" into "MobileSMS" */
function appName(bundleId: unknown): string {
  const str = String(bundleId ?? 'unknown');
  const parts = str.split('.');
  return parts[parts.length - 1] || str;
}

async function sendSpan(notif: NotifRow, content: { title?: string; body?: string; subtitle?: string }) {
  const service = appName(notif.identifier);
  const name = content.title ? `${service}: ${content.title}` : service;
  const timestampMs = (notif.delivered_date + MAC_EPOCH_OFFSET) * 1000;
  const timestampNs = String(timestampMs * 1_000_000);

  const attributes: Array<{ key: string; value: { stringValue: string } }> = [
    { key: 'service.name', value: { stringValue: service } },
  ];
  if (content.body) {
    attributes.push({ key: 'notification.body', value: { stringValue: content.body } });
  }
  if (content.subtitle) {
    attributes.push({ key: 'notification.subtitle', value: { stringValue: content.subtitle } });
  }
  attributes.push({ key: 'notification.app', value: { stringValue: notif.identifier } });

  const body = {
    resourceSpans: [{
      resource: {
        attributes: [{ key: 'service.name', value: { stringValue: service } }],
      },
      scopeSpans: [{
        spans: [{
          name,
          kind: 0, // INTERNAL
          startTimeUnixNano: timestampNs,
          endTimeUnixNano: String(timestampMs * 1_000_000 + 1_000_000), // +1ms
          status: { code: 1 }, // OK
          attributes,
        }],
      }],
    }],
  };

  try {
    const res = await fetch(`${SS_ENDPOINT}?source=notify`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${SS_API_KEY}`,
      },
      body: JSON.stringify(body),
    });
    if (res.ok) {
      const summary = [name, content.subtitle, content.body].filter(Boolean).join(' — ').slice(0, 80);
      console.log(`  ${service} / ${summary}`);
    } else {
      console.warn(`  [notif] POST failed: ${res.status}`);
    }
  } catch (err) {
    console.warn(`  [notif] POST error:`, err);
  }
}

async function poll() {
  const rows = queryNewNotifications();

  const newRows = rows.filter(r => !seenIds.has(r.rec_id));
  if (newRows.length === 0) {
    process.stdout.write('.');
    return;
  }

  console.log(`\n[notif] ${newRows.length} new notification(s)`);

  for (const row of newRows) {
    seenIds.add(row.rec_id);
    lastDeliveredDate = Math.max(lastDeliveredDate, row.delivered_date);
    const content = getNotifContent(row.rec_id);
    await sendSpan(row, content);
  }

  // Prune seen set
  if (seenIds.size > 500) {
    const arr = [...seenIds];
    seenIds.clear();
    for (const id of arr.slice(-250)) seenIds.add(id);
  }
}

// Main loop
console.log(`[notif] Polling macOS notifications every ${POLL_INTERVAL_S}s → ${SS_ENDPOINT}`);
console.log(`[notif] DB: ${DB_PATH}`);
poll();
setInterval(poll, POLL_INTERVAL_S * 1000);

File: packages/streamscapes-menubar/Sources/StreamscapesAgent/NotifPoller.swift
import Foundation

final class NotifPoller {
    private let client: OTLPClient
    private let callback: (PollerEvent) -> Void
    private let pollInterval: TimeInterval = 3
    private let macEpochOffset: Double = 978307200

    private var timer: Timer?
    private var seenIds = Set<Int>()
    private var lastDate: Double
    private var lastSuccessfulPoll: Date?
    private let dbPath: String
    private let plistPath: String

    init(endpoint: String, apiKey: String, callback: @escaping (PollerEvent) -> Void) {
        self.client = OTLPClient(endpoint: endpoint, apiKey: apiKey)
        self.callback = callback
        self.lastDate = Date().timeIntervalSince1970 - macEpochOffset
        self.dbPath = NSHomeDirectory() + "/Library/Group Containers/group.com.apple.usernoted/db2/db"
        self.plistPath = NSTemporaryDirectory() + "ss_notif_\(ProcessInfo.processInfo.processIdentifier).plist"
    }

    func start() {
        callback(.log("Notifications: polling every \(Int(pollInterval))s"))
        poll()
        timer = Timer.scheduledTimer(withTimeInterval: pollInterval, repeats: true) { [weak self] _ in
            self?.poll()
        }
    }

    func stop() {
        timer?.invalidate()
        timer = nil
    }

    func resetAfterWake() {
        lastDate = Date().timeIntervalSince1970 - macEpochOffset - 30
    }

    private func poll() {
        let query = """
        SELECT r.rec_id, a.identifier, r.delivered_date \
        FROM record r JOIN app a ON r.app_id = a.app_id \
        WHERE r.delivered_date > \(lastDate) \
        ORDER BY r.delivered_date ASC LIMIT 50;
        """

        let (output, error) = shell("sqlite3", "-json", dbPath, query)

        if let error, (error.contains("authorization denied") || error.contains("not permitted") || error.contains("unable to open")) {
            // Only flag FDA if we've never successfully read the DB
            if seenIds.isEmpty && lastSuccessfulPoll == nil {
                callback(.fdaRequired)
            }
            return
        }

        lastSuccessfulPoll = Date()

        guard let output, !output.isEmpty,
              let data = output.data(using: .utf8),
              let rows = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]
        else { return }

        let newRows = rows.filter { row in
            guard let recId = row["rec_id"] as? Int else { return false }
            return !seenIds.contains(recId)
        }
        guard !newRows.isEmpty else { return }

        callback(.log("\(newRows.count) new notifications"))

        // Cap at 8 per poll to avoid flooding the audio engine
        let capped = Array(newRows.prefix(8))

        Task {
            for (i, row) in capped.enumerated() {
                guard let recId = row["rec_id"] as? Int,
                      let identifier = row["identifier"] as? String,
                      let deliveredDate = row["delivered_date"] as? Double
                else { continue }

                seenIds.insert(recId)
                lastDate = max(lastDate, deliveredDate)

                // Stagger sends to avoid audio engine overload
                if i > 0 {
                    try? await Task.sleep(nanoseconds: 200_000_000) // 200ms
                }

                let svc = appName(identifier)
                let content = getContent(recId: recId)
                let name = content.title.map { "\(svc): \($0)" } ?? svc
                let tsMs = Int64((deliveredDate + macEpochOffset) * 1000)

                var fields: [String: Any] = ["notification.app": identifier]
                if let body = content.body { fields["notification.body"] = body }

                let result = await client.postSpan(
                    source: "notify", serviceName: svc, spanName: name,
                    fields: fields, timestampMs: tsMs
                )

                if result.ok {
                    callback(.authOk)
                    callback(.spanSent(svc))
                } else if result.status == 401 {
                    callback(.authError)
                }
            }
        }

        // Mark remaining as seen so they don't pile up
        for row in newRows.dropFirst(8) {
            if let recId = row["rec_id"] as? Int,
               let deliveredDate = row["delivered_date"] as? Double {
                seenIds.insert(recId)
                lastDate = max(lastDate, deliveredDate)
            }
        }

        // Prune
        if seenIds.count > 500 {
            seenIds = Set(Array(seenIds).suffix(250))
        }
    }

    private func appName(_ bundleId: String) -> String {
        bundleId.split(separator: ".").last.map(String.init) ?? bundleId
    }

    private func getContent(recId: Int) -> (title: String?, body: String?) {
        let _ = shell("sqlite3", dbPath,
            "SELECT writefile('\(plistPath)', data) FROM record WHERE rec_id = \(recId);")
        let (raw, _) = shell("plutil", "-p", plistPath)
        guard let raw else { return (nil, nil) }

        func extract(_ key: String) -> String? {
            let pattern = "\"\(key)\"\\s*=>\\s*\"([^\"]*)\""
            guard let regex = try? NSRegularExpression(pattern: pattern),
                  let match = regex.firstMatch(in: raw, range: NSRange(raw.startIndex..., in: raw)),
                  let range = Range(match.range(at: 1), in: raw)
            else { return nil }
            return String(raw[range])
        }

        return (extract("titl"), extract("body"))
    }

    @discardableResult
    private func shell(_ args: String...) -> (String?, String?) {
        let process = Process()
        let stdout = Pipe()
        let stderr = Pipe()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
        process.arguments = args
        process.standardOutput = stdout
        process.standardError = stderr
        do {
            try process.run()
            process.waitUntilExit()
        } catch {
            return (nil, error.localizedDescription)
        }
        let out = String(data: stdout.fileHandleForReading.readDataToEndOfFile(), encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines)
        let err = String(data: stderr.fileHandleForReading.readDataToEndOfFile(), encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines)
        return (out?.isEmpty == true ? nil : out, err?.isEmpty == true ? nil : err)
    }
}
